### PR TITLE
Try to fix #19803

### LINKED ISSUE #19803: The problem with the returned name and library parameters when requesting Type.GetType
The problem with the returned name and library parameters when requesting Type.GetType () in the event handler for the AppDomain.AssemblyResolve event.
For this, a test program was written that demonstrates the problems: https://github.com/alexger97/AssemblyResolveTest

Tested and found the problem on Linux and on Windows. Use  version Mono : 5.20.1 and .NET Framework 4.8
 

```
 AppDomain currentDomain = AppDomain.CurrentDomain;

 currentDomain.AssemblyResolve += new ResolveEventHandler(MyResolveEventHandler);

Type.GetType("CommonComponents.Infra.Zero.Uc.Deployment.ClientSoftwareUpdateStrategy, CommonComponents.Infra.Zero.Uc");
```
 ```
private static Assembly MyResolveEventHandler(object sender, ResolveEventArgs args)
        {
            Console.WriteLine($" Resolving assembly name: {args.Name}/");
            return null;
        }
```
Output in Framework:
 Resolving assembly name: CommonComponents.Infra.Zero.Uc/

Output in Mono:
 Resolving assembly name: CommonComponents.Infra.Zero.Uc, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null/

### PR DESCRIPTION
**Pull request by alexger97**
My previous unsuccessful pull request   https://github.com/mono/mono/pull/19804
I have improved the logic so that the code is more systematic

### CODE REVIEW

File: mono/metadata/assembly.c [hunk 1]
```diff
@@ -1251,13 +1251,25 @@ mono_stringify_assembly_name (MonoAssemblyName *aname)
...
+	g_string_append_printf (str, "%s%s%s", quote, aname->name, quote);
+	if (aname->has_version ==  TRUE)
+		g_string_append_printf (str,", Version=%d.%d.%d.%d", aname->major, aname->minor, aname->build, aname->revision);
+	if (aname->has_culture && aname->culture) {
+		if (*aname->culture)
+		g_string_append_printf (str, ", Culture=%s", aname->culture);
+		else
+		g_string_append_printf (str, ", Culture=%s", "neutral");
+	}
+	if (aname->has_public_key_token){
```
Comment (lambdageek) [inline mono/metadata/assembly.c:1265]:
```suggestion
	if (aname->has_public_key_token) {
```

File: mono/metadata/assembly.c [hunk 2]
```diff
@@ -1251,13 +1251,25 @@ mono_stringify_assembly_name (MonoAssemblyName *aname)
...
+		if (aname->public_key_token [0])
+			g_string_append_printf (str,", PublicKeyToken=%s%s", aname->public_key_token, (aname->flags & ASSEMBLYREF_RETARGETABLE_FLAG) ? ", Retargetable=Yes" : "");
+		else g_string_append_printf (str,", PublicKeyToken=%s%s", "null", (aname->flags & ASSEMBLYREF_RETARGETABLE_FLAG) ? ", Retargetable=Yes" : "");
+		}
```
Comment (lambdageek) [inline mono/metadata/assembly.c:1269]: Could just `if (...retargetable...) { g_string_append_printf (str, ", Retargetable=Yes"); }`  after the public key token logic - then you wouldn't have to duplicate that `?:` expression.

File: mono/metadata/reflection.c [hunk 3]
```diff
@@ -1519,6 +1521,7 @@ assembly_name_to_aname (MonoAssemblyName *assembly, char *p) {
 		*p = 0;
 		p++;
 	}
+	g_strchomp(assembly->name);
```
Comment (lambdageek) [inline mono/metadata/reflection.c:1524]:
```suggestion
	g_strchomp (assembly->name);
```

File: mono/metadata/reflection.c [hunk 4]
```diff
@@ -1549,8 +1553,9 @@ assembly_name_to_aname (MonoAssemblyName *assembly, char *p) {
 				return 1;
 			p = s;
 		} else if ((*p == 'C' || *p == 'c') && g_ascii_strncasecmp (p, "Culture=", 8) == 0) {
+			assembly->has_culture = TRUE;
 			p += 8;
-			if (g_ascii_strncasecmp (p, "neutral", 7) == 0) {
+			if ((g_ascii_strncasecmp (p, "neutral", 7) == 0) && (p [7] == ' ' || p [7] == ',')) {
```
Comment (lambdageek) [inline mono/metadata/reflection.c:1558]: Ooh... good catch!

**Comment by alexger97** [comment]
Thanks for the help, friends.
There are problems with the tests of the Corlib library, I think that I will find a solution soon. Although I have no problems with make check on Linux. I will be of any help

**Comment by lambdageek** [comment]
@alexger97 I just remembered that we also have a parallel struct in C# that we sometimes use to access a `MonoAssemblyName`:

https://github.com/mono/mono/blob/d5ed68483c83a8445737a810e7629ddd0598632e/mcs/class/corlib/Mono/RuntimeStructs.cs#L54-L71

For example here:
https://github.com/mono/mono/blob/df846bcbc9706e325f3b5dca4d09530b80e9db83/mcs/class/corlib/System.Reflection/AssemblyName.cs#L470-L534

And there are also copies in the `netcore/` directory:
https://github.com/mono/mono/blob/27921fbbd06baadeb5fa07049c587e2a214d131f/netcore/System.Private.CoreLib/src/System/Reflection/AssemblyName.Mono.cs#L63-L109
and
https://github.com/mono/mono/blob/9e00330af303bdc2b5f4ea2302e8bbcdfdc45524/netcore/System.Private.CoreLib/src/Mono/RuntimeStructs.cs#L54-L72

I'm not sure if there would be problems if you don't add your new boolean members to the managed structs, but I think it would not hurt to add the fields here too.

And with that in mind, change the field declarations in C from `gboolean` to `MonoBoolean`.  The `MonoBoolean` type is defined to interop with C# - it's a typedef for an 8-bit type same as a C# boolean; `gboolean` is a typedef for an int type which is bigger, so the layout won't match.

**Comment by alexger97** [comment]
Good afternoon ! I managed to ensure that tests that previously did not pass are now successful. However, the rest of the tests for some reason do not start.

**Comment by lambdageek** [comment]
@monojenkins build

File: mono/metadata/reflection.c [hunk 5]
```diff
@@ -1519,6 +1521,7 @@ assembly_name_to_aname (MonoAssemblyName *assembly, char *p) {
 		*p = 0;
 		p++;
 	}
+	g_strchomp (assembly->name);
```
Comment (lambdageek) [inline mono/metadata/reflection.c:1524]: Over in the https://github.com/dotnet/runtime/pull/37156 (mirror PR for the dotnet/runtime version of Mono), the builds are failing here.
```
/__w/1/s/src/mono/mono/metadata/reflection.c:1524:14: error: passing 'const char *' to parameter of type 'gchar *' (aka 'char *') discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
          g_strchomp (assembly->name);
                      ^~~~~~~~~~~~~~
  /__w/1/s/src/mono/mono/eglib/glib.h:377:39: note: passing argument to parameter 'str' here
  gchar       *g_strchomp       (gchar *str);
                                        ^
  1 error generated.
```
```suggestion
	g_strchomp ((char*)assembly->name);
```

**Comment by alexger97** [comment]
@monojenkins build

**Comment by alexger97** [comment]
Hello. Finally, the tests give a positive result at https://github.com/dotnet/runtime/pull/37156. I think you need to run all the local tests.

**Comment by alexger97** [comment]
@monojenkins build

**Comment by alexger97** [comment]
Please, need to start all test's

**Comment by thaystg** [comment]
@monojenkins build

**Comment by alexger97** [comment]
Please make all test.

**Comment by lambdageek** [comment]
@monojenkins approve

**Comment by alexger97** [comment]
https://github.com/mono/mono/pull/20152 is the first piece of suggested Mono injection code. Next, I want to focus on debugging failing tests here

**Comment by alexger97** [comment]
If this is a success, friends, then I'm very glad. I think, in this version, it could be added to master

**Comment by marek-safar** [comment]
@CoffeeFlux @lambdageek please resolve this PR

**Pull request merged**